keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1021,6 +986,50 @@ All these capabilities are required before Beta, but will not be implemented in
 the first alpha release of this enhancement due to the size and complexity of
 this enhancement.
 
+#### Informational type checking
+
+Some advantages of strongly typed objects and expressions over treating everything as unstructured are:
+
+- Checks against missing/misspelled fields.
+  The user may write an expression that refers to a missing/misspelled field that
+  does not exist in their test cases but appears later. A type check can detect this kind of error while an evaluation-time check may not.
+- Checks against type confusions. Similarly, the user may confuse the type of field but their test cases never touch wrongly typed fields.
+- Guard against short-circuit evaluation. The user may make a mistake of one of the mentioned above but the code path is never covered in their test cases;
+- Support Kubernetes extensions. For example, IntOrString and map lists.
+
+However, enforcing types for every expression and object is not feasible because of:
+
+- Version skew
+- CRDs
+- Aggregated API servers
+
+Problem examples:
+
+| Problem                                          | Summary                                                                                                               |
+|--------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------|
+| version skew: ephemeralContainers case           | New pod field, need to be able to validate in same was containers and initContainers if field exists and is populated |
+| version skew: Migration from annotation to field | Need to be able to validate annotation (if present) or field (if it exists and is populated)                          |
+| CRD is deleted                                   | Nothing to type check against, but also means there are no coresponding custom resources                              |
+| CRD is in multiple clusters, but schema differs  | If policy author is aware of the schema variations, can they write policies that work for all the variations?         |
+| Validation of an aggregated API server type      | Main API server does not have type definitions                                                                        |
+
+Because of these complications, the type checking would remain informational.
+Informational type checking will be performed against all expressions where a GVK can be resolved to
+type check against. The result of type checking will be part of the status of the performed policy.
+
+For example, accessing an unknown field will result a warning like this.
+
+```yaml
+...
+status:
+  expressionWarnings:
+    - expression: "object.replicas > 1" # should be "object.spec.replicas > 1"
+      warning: "no such field 'replicas'"
+```
+
+To address the version skew problem that can occur during upgrade/downgrade, the type checking is designed to
```

Comment:
I was trying to say it *should* be type checking and we need a real solution for upgrade/downgrade. Informational type checking doesn't seem to solve the actual problem, which is that expressions may not match schemas when the schema changes.

Reply:
I agree. Preserving schema during upgrade/downgrade is not something type checking alone can help with. Type checking is not designed to mitigate schema skews. This KEP acknowledges the upgrade/downgrade issue as part of current limitations. We will work on a solution against upgrade/downgrade independent of the type system, in another KEP update.